goalgorilla/open_social
# Issue #3163453: Issues with optional modules and their dependencies and working together.

## Problem
There are several issues we discovered when enabling all optional modules together on installation in #1899. You will see a lot of the tests failing whilst being green before, #1899 when we did not enabling these default modules. Although these modules should work together and should be able to be enabled regardless.

Most of them have to do with either not configured dependencies correctly. Usually social_core or a parent module.

Social_core contains user roles, image crop settings and other configs used through all these optional modules. 
Some optional modules are written on top of other modules (like social_group_default_route) so we need to ensure it has correct dependencies in this case social_group.
As well as issues with modules not properly co-existing within Open Social. I will explain them separately in here but treat them in one PR, since otherwise we will not be able to get the tests green to prove we are on the right track.


1. In [#3122686: Move featured content to a separate module](https://www.drupal.org/project/social/issues/3122686) we have separated the featured content module, however there are scenario's where social_page is not enabled yet when social_featured_content is being enabled (on installation mostly)
This results in

```
Drupal\Core\Config\UnmetDependenciesException: Configuration objects [error] provided by <em class="placeholder">social_featured_content</em> have unmet dependencies: <em class="placeholder">core.entity_view_display.node.page.featured (field.field.node.page.body, field.field.node.page.field_content_visibility, field.field.node.page.field_files, field.field.node.page.field_page_comments, field.field.node.page.field_page_image, node.type.page)</em>
```

2. A lot of the optional modules don't have social_core or a parent as their dependency which results in them being disabled without for example user roles being created. This results in install hooks firing, trying to add permissions to a SM/CM and these roles are not there yet. I have added some examples I encountered and after this PR they should turn greena gain.

3. Group has different ways of making sure the `entity.group.canonical` goes to it's respective page. Which by default is the stream. For Closed groups it's the about page. Unless we have The Social Group Default Route installed, which then takes over. Also the Social path manager has some redirects in there. Basically all of these are now enabled and we need them to co-exist, therefor issues with not being able to join a group or the group stream not working were encountered. 

4. Social profile privacy could potentially be enabled before social_search, if that is the case the install hook which allows for the processor that restricts usernames is not added correctly on install. Therefor we have added the social_search as dependency to ensure this install hooks works well.

5. There seemed to be an issue with AN viewing public posts.

## How to test
Just re-install open_social and include all the optional modules to see it still works, but above all make sure all tests are green again. 

## Screenshots
1. Moving it to the optional module should fix it.
 ![Screenshot 2020-08-05 at 11 32 41](https://user-images.githubusercontent.com/16667281/89396914-75429300-d70f-11ea-9124-9036a93488ae.png)

2. You see here how GDPR didn't work, due to the social_core module not being a dependency permissions were not set for CM/SM when installing the module.

<img width="1211" alt="Screenshot 2020-08-08 at 12 20 55" src="https://user-images.githubusercontent.com/16667281/89707784-a9b58980-d971-11ea-9b3e-e18b78542146.png">


3. You see here the stream is not working so people can't join.

<img width="1372" alt="Bugfix social default route combi with paths part 2" src="https://user-images.githubusercontent.com/16667281/89707775-8ab6f780-d971-11ea-9d6b-5b2bc880e835.png">

4. The first name last name nick name restrcition doesn't work correctly.

![Screenshot 2020-08-08 at 13 06 44](https://user-images.githubusercontent.com/16667281/89708710-bb9b2a80-d979-11ea-991d-0c2d6d1048a9.png)
![Screenshot 2020-08-08 at 13 12 56](https://user-images.githubusercontent.com/16667281/89708711-be961b00-d979-11ea-904e-78a6968d9b4d.png)

5. There seemed to be an issue with AN viewing public posts.

## Release notes
All the optional modules enabled in #1899 should now work fine on install.
